src/hal/src/command/render_pass.rs

```diff
@@ -156,6 +159,34 @@ impl<'a, B: Backend> RenderSubpassCommon<'a, B> {
     // TODO: begin/end query
 }
 
+fn convert_begin_render_pass_clear_values<T>(clear_values: T) -> impl Iterator<Item=ClearValueRaw>
```

Comment:
This is unfortunate. It would be nice to not have this piece of logic between typed gfx-hal and raw one

Reply:
yeah, maybe inside each backend? As I see backends enumerate over clear values anyways.

Reply:
Can it go into https://github.com/gfx-rs/gfx/tree/master/src/backend/auxil ?

Reply:
yes, but if it goes there, each backend calls it, is that right?

Reply:
yes. Also, we should make it work on an existing `Vec` rather than allocating one

Reply:
I'm not sure how to make it work on an existing `Vec`. I have to sort the clear values which can be only done on the incoming clear values, not on the existing `Vec`s. Currently it uses `SmallVec`.

Reply:
the user would provide you with an iterator and a `&mut Vec<>` for the output

Reply:
By user you mean the caller? How would an output `Vec` help as a parameter? As I see the only difference is that caller would have to do the allocation instead of the callee, but it's still one extra allocation.

Reply:
The caller can choose to pre-allocate and re-use the same vector many times to avoid all but a single allocation.

Reply:
I can place the `Vec` inside the `CommandBuffer` struct, so it can be reused, but from a maintainability perspective, it doesn't sound the best to place every `Vec` inside `CommandBuffer`.

Reply:
I think this complication is rather unfortunate, and it's only caused by the fact that you want to include attachment indices in there in the first place. We don't have to, and without indices there is no need to collect anything on the backend/hal side or sort.

Let's get back to the original idea and just pass the clear colors with expected ordering.

Reply:
What was the original idea?

The simplest way seems to be the one that Vulkan is using now, an array of clear values that has the values at the same index as the attachment indices.

If the array was only containing the clear values for attachments with a clearop, then you would have to track the attachments in the vulkan hal backend to readd the dummy values. Also portability would have a hard time filtering for those clear values that are for attachments with clearop.

Reply:
The original idea is to have an iterator of clear values that are only for the attachments that need clears. The current API has a clear value per attachment, and not only some of them are unused, it may have less total values than attachments. This is unsound.
We need to either have strictly a value per attachment, or (better) a value per cleared attachment.

Reply:
The better solution is a bigger work in my opinion, but someone can do it in a new PR.

Having a value per attachment, is almost the same as it is now, only need to check if the iterator has the same length as the bound render pass in the command buffer. This can be also complicated for portability, because it has to know how many attachments are there when calling `begin_render_pass`.

Reply:
> The better solution is a bigger work in my opinion, but someone can do it in a new PR.

Why do you think so?

> This can be also complicated for portability, because it has to know how many attachments are there when calling begin_render_pass

It's not a big stretch for the render pass to know how many attachments it has, at the portability level.

Reply:
Sorry, I just realized that my explanation was a flawed, I'm tired a bit.

So these are the things that needs to be done to get there if I understand right:

The render pass needs to be extended on the vulkan backend to know what attachments are there. Other backends already have this information as far as I know. And the portability layer needs to be able to query the same information about the render pass's attachments. After these are done, it's easy to implement the clear values per cleared attachments.